akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -210,6 +210,166 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
+    parent ⇒
+    lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider).getOrElse(Supervision.stoppingDecider)
+    private val activeSubstreamsMap = mutable.Map.empty[Any, SubstreamSource]
+    private val closedSubstreams = mutable.Set[Any]()
+    private var timeout: FiniteDuration = _
+    private var substreamWaitingToBePushed: Option[SubstreamSource] = None
+    private var nextElement: (K, T) = null
+    private var _nextId = 0
+
+    private def nextId(): Long = { _nextId += 1; _nextId }
+
+    private def hasNextElement = nextElement ne null
+
+    private def completeWholeStage(): Boolean =
+      if (activeSubstreamsMap.isEmpty || (!hasNextElement && !activeSubstreamsMap.values.exists(_.firstPush))) {
+        activeSubstreamsMap.values.foreach(_.complete())
+        completeStage()
+        true
+      } else false
+
+    private def fail(ex: Throwable): Unit = {
+      activeSubstreamsMap.values.foreach(_.fail(ex))
+      failStage(ex)
+    }
+
+    private def needToPull: Boolean = !(hasBeenPulled(in) || isClosed(in) || hasNextElement)
+
+    override def preStart(): Unit = timeout = ActorMaterializer.downcast(interpreter.materializer).settings.subscriptionTimeoutSettings.timeout
+
+    override def onPull(): Unit = {
+      substreamWaitingToBePushed match {
+        case Some(substreamSource) ⇒
+          push(out, Source.fromGraph(substreamSource.source))
+          scheduleOnce(substreamSource.key, timeout)
+          substreamWaitingToBePushed = None
+        case None ⇒
+          if (hasNextElement) {
+            activeSubstreamsMap(nextElement._1).push(nextElement._2)
+            nextElement = null
+          } else tryPull(in)
+      }
+    }
+
+    override def onUpstreamFailure(ex: Throwable): Unit = fail(ex)
+
+    override def onDownstreamFinish(): Unit =
+      if (activeSubstreamsMap.isEmpty) completeStage() else setKeepGoing(true)
+
+    override def onPush(): Unit = {
+      val elem = grab(in)
+      val key: K = try {
+        val k = f(elem)
+        require(k != null, "Key cannot be null")
+        k
+      } catch {
+        case NonFatal(ex) ⇒
+          decider(ex) match {
+            case Supervision.Stop                         ⇒ fail(ex)
+            case Supervision.Resume | Supervision.Restart ⇒ if (!hasBeenPulled(in)) pull(in)
+          }
+          null.asInstanceOf[K]
+      }
+
+      if (key != null) {
+        activeSubstreamsMap.get(key) match {
+          case Some(substreamSource) ⇒
+            if (substreamSource.isAvailable) substreamSource.push(elem)
+            else nextElement = key -> elem
+          case None ⇒
+            if (activeSubstreamsMap.size == maxSubstreams)
+              fail(new IllegalStateException(s"Cannot open substream for key '$key': too many substreams open"))
+            else if (closedSubstreams(key) && !hasBeenPulled(in)) pull(in)
+            else runSubstream(key, elem)
+        }
```

Comment:
This block could be on line 273 instead of catching the value of the try-catch and then if:ing, if it key was null it will never reach there and if there was an exception it will always be null anyway.